cpp/include/raft/matrix/detail/scatter.cuh

```diff
@@ -0,0 +1,127 @@
+ * into the appropriate location in the scratch space and copied back to
+ * the corresponding locations in the input matrix.
+ *
+ * @tparam InputIteratorT
+ * @tparam MapIteratorT
+ * @tparam IndexT
+ *
+ * @param[inout] handle raft handle
+ * @param[inout] inout input matrix (n_rows * n_cols)
+ * @param[inout] map map containing the destination index for each row (n_rows)
+ * @param[inout] batch_size column batch size
+ */
+
+template <typename MatrixT, typename IndexT>
+void scatterInplaceImpl(
+  raft::resources const& handle,
+  raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+  raft::device_vector_view<const IndexT, IndexT, raft::layout_c_contiguous> map,
+  IndexT batch_size)
+{
+  IndexT m          = inout.extent(0);
+  IndexT n          = inout.extent(1);
+  IndexT map_length = map.extent(0);
+
+  // skip in case of 0 length input
+  if (map_length <= 0 || m <= 0 || n <= 0 || batch_size < 0) return;
+
+  RAFT_EXPECTS(map_length == m,
+               "Length of map should be equal to number of rows for inplace scatter");
+
+  // re-assign batch_size for default case
+  if (batch_size == 0) batch_size = n;
+
+  RAFT_EXPECTS(batch_size <= n, "batch size should be <= number of columns");
+
+  auto exec_policy = resource::get_thrust_policy(handle);
+
+  IndexT n_batches = raft::ceildiv(n, batch_size);
+
+  auto scratch_space = raft::make_device_vector<MatrixT, IndexT>(handle, m * batch_size);
+
+  for (IndexT bid = 0; bid < n_batches; bid++) {
+    IndexT batch_offset   = bid * batch_size;
+    IndexT cols_per_batch = min(batch_size, n - batch_offset);
+
+    auto copy_op = [inout = inout.data_handle(),
+                    map   = map.data_handle(),
+                    batch_offset,
+                    cols_per_batch = raft::util::FastIntDiv(cols_per_batch),
+                    n] __device__(auto idx) {
+      IndexT row = idx / cols_per_batch;
+      IndexT col = idx % cols_per_batch;
+      return inout[row * n + batch_offset + col];
+    };
+    raft::linalg::map_offset(
+      handle,
+      raft::make_device_vector_view(scratch_space.data_handle(), m * cols_per_batch),
+      copy_op);
+
+    auto scatter_op = [inout         = inout.data_handle(),
+                       map           = map.data_handle(),
+                       scratch_space = scratch_space.data_handle(),
+                       batch_offset,
+                       cols_per_batch = raft::util::FastIntDiv(cols_per_batch),
+                       n] __device__(auto idx) {
+      IndexT row     = idx / cols_per_batch;
+      IndexT col     = idx % cols_per_batch;
+      IndexT map_val = map[row];
+
+      inout[map_val * n + batch_offset + col] = scratch_space[idx];
+      return;
+    };
+    auto counting = thrust::make_counting_iterator<IndexT>(0);
+    thrust::for_each(exec_policy, counting, counting + m * cols_per_batch, scatter_op);
+  }
+}
+
+template <typename MatrixT, typename IndexT>
+void scatter(raft::resources const& handle,
```

Comment:
Simply decided to overload the current gather and scatter because the set of args is different and this comment from Corey. https://github.com/rapidsai/raft/pull/1445#discussion_r1194390748

Reply:
I followed up there https://github.com/rapidsai/raft/pull/1445#discussion_r1227055158